File: imperative/java-imperative/src/main/java/com/fernandoglizt/imperative/javaimperative/service/ImperativeProcessor.java
package com.fernandoglizt.imperative.javaimperative.service;

import com.fernandoglizt.imperative.javaimperative.model.ProcessRequest;
import com.fernandoglizt.imperative.javaimperative.model.ProcessResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

@Service
public class ImperativeProcessor {

    private static final Logger log = LoggerFactory.getLogger(ImperativeProcessor.class);

    private final RestTemplate restTemplate;
    private final ExecutorService executorService;
    
    @Value("${app.downstream.retry-attempts:1}")
    private int retryAttempts;

    public ImperativeProcessor(RestTemplate restTemplate, ExecutorService executorService) {
        this.restTemplate = restTemplate;
        this.executorService = executorService;
    }

    public ProcessResponse processEvents(ProcessRequest request) {
        long startTime = System.currentTimeMillis();
        
        List<Integer> items = IntStream.rangeClosed(1, request.count())
            .map(i -> i * 2)
            .filter(i -> i % 2 == 0)
            .boxed()
            .toList();

        AtomicInteger externalCallsOk = new AtomicInteger(0);
        AtomicInteger externalCallsFail = new AtomicInteger(0);
        
        List<List<Integer>> batches = createBatches(items, request.batch());
        int batchCount = batches.size();
        
        CountDownLatch latch = new CountDownLatch(items.size());
        
        for (List<Integer> batch : batches) {
            for (Integer item : batch) {
                executorService.submit(() -> {
                    try {
                        boolean success = callDownstreamWithRetry(
                            item, 
                            request.getIoDelayMs(), 
                            request.getDownstreamUrl()
                        );
                        if (success) {
                            externalCallsOk.incrementAndGet();
                        } else {
                            externalCallsFail.incrementAndGet();
                        }
                    } finally {
                        latch.countDown();
                    }
                });
            }
        }

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            log.error("Processing interrupted", e);
        }

        long durationMs = System.currentTimeMillis() - startTime;
        String hostname = getHostname();

        ProcessResponse response = ProcessResponse.success(
            request,
            items.size(),
            batchCount,
            externalCallsOk.get(),
            externalCallsFail.get(),
            durationMs,
            hostname
        );

        log.info("Processed {} events in {}ms ({} events/sec) - OK: {}, Fail: {}",
            items.size(), durationMs, response.eventsPerSec(),
            externalCallsOk.get(), externalCallsFail.get());

        return response;
    }

    private boolean callDownstreamWithRetry(int item, int delayMs, String baseUrl) {
        String url = appendDelayParam(baseUrl, delayMs);
        
        for (int attempt = 0; attempt <= retryAttempts; attempt++) {
            try {
                restTemplate.getForObject(url, String.class);
                return true;
            } catch (Exception e) {
                if (attempt < retryAttempts) {
                    try {
                        long backoff = Math.min(100L * (1L << attempt), 300L);
                        Thread.sleep(backoff);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        return false;
                    }
                }
            }
        }
        return false;
    }

    private String appendDelayParam(String url, int delay) {
        return url.contains("?") 
            ? url + "&delay_ms=" + delay 
            : url + "?delay_ms=" + delay;
    }

    private List<List<Integer>> createBatches(List<Integer> items, int batchSize) {
        List<List<Integer>> batches = new ArrayList<>();
        for (int i = 0; i < items.size(); i += batchSize) {
            batches.add(items.subList(i, Math.min(i + batchSize, items.size())));
        }
        return batches;
    }

    private String getHostname() {
        try {
            return InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            return "unknown";
        }
    }
}

File: imperative/kotlin-imperative/src/main/kotlin/com/fernandoglizt/imperative/kotlinimperative/service/ImperativeProcessor.kt
package com.fernandoglizt.imperative.kotlinimperative.service

import com.fernandoglizt.imperative.kotlinimperative.model.ProcessRequest
import com.fernandoglizt.imperative.kotlinimperative.model.ProcessResponse
import kotlinx.coroutines.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.net.InetAddress
import java.util.concurrent.atomic.AtomicInteger

@Service
class ImperativeProcessor(
    private val restTemplate: RestTemplate
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Value("\${app.downstream.retry-attempts:1}")
    private val retryAttempts: Int = 1

    @Value("\${app.concurrency:128}")
    private val concurrency: Int = 128

    suspend fun processEvents(request: ProcessRequest): ProcessResponse = coroutineScope {
        val startTime = System.currentTimeMillis()

        val items = (1..request.count)
            .map { it * 2 }
            .filter { it % 2 == 0 }

        val externalCallsOk = AtomicInteger(0)
        val externalCallsFail = AtomicInteger(0)

        val batches = items.chunked(request.batch)
        val batchCount = batches.size

        // Process all items with controlled concurrency
        items.chunked(concurrency).forEach { chunk ->
            val jobs = chunk.map { item ->
                async(Dispatchers.IO) {
                    val success = callDownstreamWithRetry(
                        item,
                        request.getIoDelayMsNormalized(),
                        request.getDownstreamUrlOrDefault()
                    )
                    if (success) {
                        externalCallsOk.incrementAndGet()
                    } else {
                        externalCallsFail.incrementAndGet()
                    }
                }
            }
            jobs.awaitAll()
        }

        val durationMs = System.currentTimeMillis() - startTime
        val hostname = getHostname()

        val response = ProcessResponse.success(
            request = request,
            processedEvents = items.size,
            batches = batchCount,
            externalCallsOk = externalCallsOk.get(),
            externalCallsFail = externalCallsFail.get(),
            durationMs = durationMs,
            hostname = hostname
        )

        log.info(
            "Processed {} events in {}ms ({} events/sec) - OK: {}, Fail: {}",
            items.size, durationMs, response.eventsPerSec,
            externalCallsOk.get(), externalCallsFail.get()
        )

        response
    }

    private suspend fun callDownstreamWithRetry(item: Int, delayMs: Int, baseUrl: String): Boolean {
        val url = appendDelayParam(baseUrl, delayMs)

        repeat(retryAttempts + 1) { attempt ->
            try {
                // Call blocking RestTemplate from IO dispatcher
                withContext(Dispatchers.IO) {
                    restTemplate.getForObject(url, String::class.java)
                }
                return true
            } catch (e: Exception) {
                if (attempt < retryAttempts) {
                    val backoff = minOf(100L * (1L shl attempt), 300L)
                    delay(backoff)
                }
            }
        }
        return false
    }

    private fun appendDelayParam(url: String, delay: Int): String {
        return if ('?' in url) "$url&delay_ms=$delay" else "$url?delay_ms=$delay"
    }

    private fun getHostname(): String {
        return try {
            InetAddress.getLocalHost().hostName
        } catch (e: Exception) {
            "unknown"
        }
    }
}

